liquid-compiler/src/parser.rs

```diff
@@ -3,622 +3,411 @@
+            .expect("A filterchain always has starts by a value."),
+    );
+    let filters = chain.map(parse_filter).collect();
+
+    FilterChain::new(entry, filters)
 }
 
-/// Recognises a value token, returning an error if a non-value token
-/// is presented.
-pub fn value_token(t: Token) -> Result<Token> {
-    match t {
-        v @ Token::StringLiteral(_)
-        | v @ Token::IntegerLiteral(_)
-        | v @ Token::FloatLiteral(_)
-        | v @ Token::BooleanLiteral(_)
-        | v @ Token::Identifier(_) => Ok(v),
-        x => Err(unexpected_token_error(
-            "string | number | boolean | identifier",
-            Some(&x),
-        )),
+fn parse_tag(
+    tag: Pair,
+    next_elements: &mut Iterator<Item = Pair>,
+    options: &LiquidOptions,
+) -> Result<Box<Renderable>> {
+    if tag.as_rule() != Rule::Tag {
+        panic!("Expected a tag.");
     }
-}
 
-/// Describes the optional trailing part of a block split.
-pub struct BlockSplit<'a> {
-    pub delimiter: String,
-    pub args: &'a [Token],
-    pub trailing: &'a [Element],
+    let mut tag = tag.into_inner();
+    let name = tag
+        .next()
+        .expect("A tag starts with an identifier.")
+        .as_str();
+    let tokens = TagTokens(&mut tag);
+
+    if options.tags.contains_key(name) {
+        options.tags[name].parse(name, tokens, options)
+    } else if options.blocks.contains_key(name) {
+        parse_block(name, tokens, next_elements, options)
+    } else {
+        panic!("Errors not implemented. Unknown tag.")
+    }
 }
 
-/// A sub-block aware splitter that will only split the token stream
-/// when it finds a delimter at the top level of the token stream,
-/// ignoring any it finds in nested blocks.
-///
-/// Returns a slice contaiing all elements before the delimiter, and
-/// an optional `BlockSplit` struct describing the delimiter and
-/// trailing elements.
-pub fn split_block<'a>(
-    tokens: &'a [Element],
-    delimiters: &[&str],
+fn parse_block(
+    name: &str,
+    tokens: TagTokens,
+    next_elements: &mut Iterator<Item = Pair>,
     options: &LiquidOptions,
-) -> (&'a [Element], Option<BlockSplit<'a>>) {
-    // construct a fast-lookup cache of the delimiters, as we're going to be
-    // consulting the delimiter list a *lot*.
-    let delims: HashSet<&str> = HashSet::from_iter(delimiters.iter().map(|x| *x));
-    let mut stack: Vec<String> = Vec::new();
-
-    for (i, t) in tokens.iter().enumerate() {
-        if let Element::Tag(ref args, _) = *t {
-            match args[0] {
-                Token::Identifier(ref name) if options.blocks.contains_key(name.as_str()) => {
-                    stack.push("end".to_owned() + name);
-                }
+) -> Result<Box<Renderable>> {
+        let mut nesting_depth = 1;
+        let mut block = TagBlock::new();
 
-                Token::Identifier(ref name) if Some(name) == stack.last() => {
-                    stack.pop();
+        while let Some(element) = next_elements.next() {
```

Comment:
To be clear, I feel like we could implement this nom-like approach in Pest.   As long as it gives us where the parser left off, we could get this going.  We'd have to pass along that information as well as line / column information since pest wouldn't realize we were picking up again within the same file.

Reply:
I'm thinking that maybe `TagBlock` could be wrapping the iterator of the whole "stream of elements" and its functions (that would be used in block tags creation) would silently check if it is parsing the close tag, stopping at that point.

Reply:
I made some changes to the `TagBlock`. Now, the check for the end tag occurs inside of its `next` function (in other words, when a new element is being requested from the "stream of elements"). At the end, the `TagBlock` closes, leaving the iterator at the right place for the next element to be parsed. 

Is this the type of approach desired?

Reply:
We'll see how it works.